### PR TITLE
docs: Updated the Beaker 1.0 migration documentation

### PR DESCRIPTION
**Pull request by robdmoore**

### CODE REVIEW

File: docs/source/migration.rst [hunk 1]
```diff
@@ -85,7 +108,7 @@ The arguments that ``Application.__init__()`` takes have changed, the following
         optimize_options=pyteal.OptimizeOptions(scratch_slots=False, frame_pointers=False),
     )
 
-becomes this in ``1.0``:
+the becomes the following in ``1.0``:
```
Comment (achidlow) [inline docs/source/migration.rst:111]: “then”?

File: docs/source/migration.rst [hunk 2]
```diff
@@ -1,7 +1,31 @@
...
+We have carefully articulated all considerations in this document that should be evaluated to perform a
+successful migration of a ``0.x`` Beaker contract/signature to a ``1.0`` compatible one.
+
+While there are a number of considerations to work through, care was taken to ensure that the migration
+effort should be relatively straightforward. If you have any problems or questions about a migration feel
+free to send messages to the `algokit channel in Algorand Discord <https://discord.com/channels/491256308461207573/1065320801970180168>`_.
+
+In order to de-risk the migration effort we recommend you test for output stability of the TEAL code
+that is generated from your Beaker code. This involves storing the current TEAL and application spec output from calling
+``.dump()`` on your Beaker contract and then doing a diff after performing the migration. We expect there will be
```
Comment (daniel-makerx) [inline docs/source/migration.rst:22]: .export() now, see Application.compile() section

File: docs/source/migration.rst [hunk 3]
```diff
@@ -547,24 +572,24 @@ in ``1.0`` this becomes:
...
+Precompiled signatures and applications
+---------------------------------------
 
 In ``0.x`` logic signatures and applications could be precompiled by adding an ``AppPrecompile`` or
 ``LSigPrecompile`` attribute to the application class, making certain properties available for use inside
-the application's methods.
+the application's methods like the logic hash and the TEAL code.
 
-In ``1.0``, you do not need to reference any "precompile classes" directly, instead use the ``beaker.precompiled`` function.
+In ``1.0``, you do not need to reference any "precompile classes" directly, instead you can use the ``beaker.precompiled`` function.
```
Comment (achidlow) [inline docs/source/migration.rst:590]: It’s not optional, which is why I avoided language like “you can”

File: docs/source/migration.rst (hunk 2 above)
Comment (daniel-makerx) [reply docs/source/migration.rst:22]: Actually to get a diff, it is ``.dump()`` on the original ``0.x`` application and then ``.export()`` on the specification returned from ``.build()`` after migration to ``1.0``

**Comment by barnjamin** [comment]
> Previously, this was considered valid PyTeal, however since the clear state program can not reject, there is no way to ensure these arguments are available, leading to silent failures.

slight nit, they _can_ reject but doing so _might_ cause issues with your app if there is certain cleanup or book-keeping logic you need to execute

**Comment by barnjamin** [comment]
https://github.com/algorand-devrel/beaker/blob/91b7e98df8c340fd6cde85722bd0f66c70489897/docs/source/migration.rst#signer

Why remove util method to get signer from a precompile?

**Comment by barnjamin** [comment]
> In addition, the interface of precompiled logic signature objects has been simplified. As can be seen in the example above, obtaining the address is done via .address() instead of .logic.hash() for normal logic signatures.

for apps, it is often useful to get the hash of the program (or of each page) and it would _not_ be the address, has that been taken out?

**Comment by achidlow** [comment]
@barnjamin 

> slight nit, they can reject but doing so might cause issues with your app if there is certain cleanup or book-keeping logic you need to execute

How should we rephrase this to say they can't stop the transaction from being accepted anyway?

> Why remove util method to get signer from a precompile?

This is part of trying to split `ApplicationClient` & friends out of beaker - and also attempting to make the Application Specification "language agnostic". It's not 100% there as it stands, due to these precompiles. We would need to add LSigs + Runtime Templated LSigs into the app-spec, and also standardise the Runtime Templated LSigs so a `populate_template(...)` could be implemented in other languages. I'm open to adding them back in, since I didn't entirely achieve the decoupling I was aiming for.

> for apps, it is often useful to get the hash of the program (or of each page) and it would not be the address, has that been taken out?

So this looked like an internal implementation detail to me, and was able to be refactored away. If there's a use case for wanting those values from a precompiled `Application` within PyTeal generating code, we can definitely add it back in?

**Comment by achidlow** [comment]
ping @barnjamin , in case the above got lost due to this PR being merged/closed :)

**Comment by barnjamin** [comment]
current
> however since the clear state program can not reject, there is no way to ensure these arguments are available, leading to silent failures.

propsed
> however since a clear state program `reject` will not prevent the accounts local state from being cleared, special care needs to be taken to allow as few conditions that might lead to rejection as possible. 

as for the other 2 things, I'm ok leaving them out until someone asks for them, I like all the clean up work and don't love the idea of adding back features that _maybe_ someone was working.

"crank it til it strips then back off a quarter turn"
